Soci snapshotter change service

ginglis13 opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #580

**Description of changes:**

Bottlerocket will allow users to modify the snapshotter configured for containerd via settings.container-runtime.snapshotter. Switching snapshotters requires the clean state of containerd, meaning no active
running containers or existing images. Bottlerocket should only make the change to configured snapshotter at boot of an instance and not do so at runtime.

This commit  adds a service to configure snapshotter selection change on reboot of an instance. When settings.container-runtime.snapshotter is modified, it will only take effect at reboot. Change in containerd's underlying snapshotter can lead to dangling content, so when such a change is detected, reset respective snapshotter state directories /var/lib/containerd and /var/lib/soci-snapshotter-grpc.

**Testing done:**

In each case, launch the same pod listed below. Chosen variant: `aws-k8s-1.33-x86_64`.

<details>

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: djl-deployment
  labels:
    app: djl-soci
spec:
  replicas: 1
  selector:
    matchLabels:
      app: djl-soci
  template:
    metadata:
      labels:
        app: djl-soci
    spec:
      containers:
      - name: djl-container
        #image: deepjavalibrary/djl-serving:0.26.0-deepspeed
        image: 763104351884.dkr.ecr.us-west-2.amazonaws.com/djl-inference:0.22.1-deepspeed0.9.2-cu118
        command: ["sleep"]
        args: ["inf"]
```

</details>

- [x] Reboot an instance with no change in snapshotter selection
  * 1x for soci and 1x for overlay configured
  * ensure all pods come back up
- [x] Boot with soci. Run some pods with example images. change snapshotter. reboot
  * ensure pods come back up to `Running`
  * ensure `ls -lR /var/lib/soci-snapshotter` is empty
- [x] Boot with overlay. Run some pods. change snapshotter. reboot.
  * ensure pods come back up to `Running`
  * ensure content via `ls -lR /var/lib/soci-snapshotter`
* **RESULTS: https://gist.github.com/ginglis13/cfc7d1d907e65d889af9b55923476e12**

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

ginglis13 committed:
release: add service to configure snapshotters on reboot

Add service to configure snapshotter selection change on reboot of an
instance. When settings.container-runtime.snapshotter is modified, it
will only take effect at reboot. Change in containerd's underlying
snapshotter can lead to dangling content, so when such a change is
detected, reset respective snapshotter state directories
/var/lib/containerd and /var/lib/soci-snapshotter.

Require soci-snapshotter only on k8s variants.

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

ginglis13 committed:
containerd-1.7: start service after snapshotters configured

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

ginglis13 committed:
containerd-2.0: start service after snapshotters configured

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

ginglis13 committed:
selinux-policy: allow trusted_s to mutate cache_t files

In order for the snapshotter configuration service to be able to clean
up snapshotter state directories via "find", processes given trusted_s
must be allowed to mutate files that are cache_t, which is given to
containerd's state directory.

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

arnaldo2792 commented on packages/os/configure-snapshotter.service:
```diff
@@ -0,0 +1,28 @@
+ExecCondition=[ -n "${SELECTED_SNAPSHOTTER}" ]
+# Check if the active snapshotter has changed
+ExecCondition=[ "${SELECTED_SNAPSHOTTER}" != "${ACTIVE_SNAPSHOTTER}" ]
+ExecStart=find /var/lib/soci-snapshotter-grpc -mindepth 1 -delete -true
+ExecStart=find /var/lib/containerd -mindepth 1 -delete -true
```
For commands like this, use the full path to the binary.

arnaldo2792 commented on packages/os/os.spec:
```diff
@@ -424,6 +428,11 @@ Conflicts: %{_cross_os}bash
 
+%package -n %{_cross_os}configure-snapshotter
+Summary: Configure snapshotter service
+%description -n %{_cross_os}configure-snapshotter
+Service to configure snapshotter and clean up containerd and soci-snapshotter-grpc directories when snapshotter changes.
```
It is convention in bottlerocket to use the `summary` as the description of the package.

arnaldo2792 commented on packages/os/os.spec:
```diff
@@ -89,6 +92,7 @@ Requires: %{_cross_os}storewolf
 Requires: %{_cross_os}sundog
 Requires: %{_cross_os}xfscli
 Requires: %{_cross_os}thar-be-settings
+Requires: %{_cross_os}configure-snapshotter
```
Are you adding this here because the snapshotter is independent of the SOCI package or containerd? If that's the case, this will be better at `release.spec` instead of `os.spec`.

`os.spec` is mostly used to compile first party Rust. `release.spec` is where we put things that should exist in all variants.

bcressey replied:
+1, definitely put this in `release.spec`. I cringe whenever anything triggers a rebuild for `os`.

arnaldo2792 commented on packages/os/os.spec:
```diff
@@ -424,6 +428,11 @@ Conflicts: %{_cross_os}bash
 %{summary}.
 
+%package -n %{_cross_os}configure-snapshotter
+Summary: Configure snapshotter service
```
```suggestion
Summary: Service and configuration files to set the containerd active snapshotter
```

arnaldo2792 commented on packages/containerd-1.7/containerd.service:
```diff
@@ -3,6 +3,7 @@ Description=containerd container runtime
 Documentation=https://containerd.io
 After=network-online.target configured.target
 Wants=network-online.target configured.target
+Requires=snapshotters-configured.service
```
If you want containerd to start after `snapshotters-configured.servce`, you should add the service to the `After=` of this unit.

Also the name of your service is wrong: `configure-snapshotter.service` is what you used in the first commit.

arnaldo2792 commented on packages/containerd-2.0/containerd.service:
```diff
@@ -3,6 +3,7 @@ Description=containerd container runtime
 Documentation=https://containerd.io
 After=network-online.target configured.target
 Wants=network-online.target configured.target
+Requires=snapshotters-configured.service
```
Same as elsewhere regarding ordering and the name of the service.

arnaldo2792 commented on packages/selinux-policy/rules.cil:
```diff
@@ -246,6 +246,7 @@
 (allow clock_s measure_t (files (mutate)))
```
For the lack of a better place to put this, your commit message refers to a label that doesn't exist (`process_s`). I think you meant `trusted_s`.

bcressey commented on packages/selinux-policy/rules.cil:
```diff
@@ -246,6 +246,7 @@
 (allow clock_s measure_t (files (mutate)))
 (allow network_s lease_t (files (mutate)))
 (allow runtime_s cache_t (files (mutate)))
+(allow trusted_s cache_t (files (mutate)))
```
You can change the line above from `runtime_s` to `trusted_s` (which includes `runtime_s`).

bcressey commented on packages/os/configure-snapshotter.service:
```diff
@@ -0,0 +1,28 @@
+
+[Service]
+Type=oneshot
+# This file might not exist as users don't have to set `settings.container-runtime.snapshotter`
+EnvironmentFile=-/etc/selected-snapshotter
```
Let's put this under `/etc/containerd/selected-snapshotter`.

bcressey commented on packages/os/configure-snapshotter.service:
```diff
@@ -0,0 +1,28 @@
+ExecCondition=[ -n "${SELECTED_SNAPSHOTTER}" ]
+# Check if the active snapshotter has changed
+ExecCondition=[ "${SELECTED_SNAPSHOTTER}" != "${ACTIVE_SNAPSHOTTER}" ]
+# Don't error if /var/lib/soci-snapshotter doesn't exist, which it may not if soci is not selected as snapshotter.
+ExecStart=-/x86_64-bottlerocket-linux-gnu/sys-root/usr/bin/find /var/lib/soci-snapshotter -mindepth 1 -delete -true
+ExecStart=-/x86_64-bottlerocket-linux-gnu/sys-root/usr/bin/find /var/lib/containerd -mindepth 1 -delete -true
```
Need to remove the sysroot prefix or else it won't work on aarch64:
```suggestion
# Don't error if the directories don't exist.
ExecStart=-/usr/bin/find /var/lib/soci-snapshotter /var/lib/containerd -mindepth 1 -delete -true
```

bcressey commented on packages/os/configure-snapshotter.service:
```diff
@@ -0,0 +1,28 @@
+[Unit]
+Description=Configure Snapshotter
+Before=containerd.service
+After=settings-applier.service bootstrap-commands.service
+Requires=settings-applier.service bootstrap-commands.service
```
You don't need these dependencies since you're running in `multi-user.target`; they'll be completed by the time this unit is queued to start.

bcressey commented on packages/os/configure-snapshotter.service:
```diff
@@ -0,0 +1,28 @@
+Type=oneshot
+# This file might not exist as users don't have to set `settings.container-runtime.snapshotter`
+EnvironmentFile=-/etc/selected-snapshotter
+# Don't error if this doesn't exist yet, which it won't on first boot.
+EnvironmentFile=-/var/cache/containerd/active-snapshotter
```
nit: not really true since we create it via `tmpfiles`

ginglis13 commented:
^ force pushes addresses feedback:

* move `configure-snapshotter.service` to the `release` package
* update some incorrect comments
* put selected snapshotter under /etc/containerd/selected-snapshotter
* remove sysroot prefix from `find` absolute path
* remove the dependencies `settings-applier.service` and `bootstrap-commands.service` from `configure-snapshotter.service`
* replace with `runtime_s cache_t` rule with `trusted_s cache_t` rule given `trusted_s` includes `runtime_s`


Additionally, provide `soci-snapshotter` in `release` if variant is k8s.

bcressey commented on packages/release/release.spec:
```diff
@@ -149,6 +154,11 @@ Requires: %{_cross_os}libkcapi
+Summary: Service and configuration files to set the containerd active snapshotter
+%description -n %{_cross_os}configure-snapshotter
+%{summary}.
+
```
I don't think we need a separate subpackage for this.

ginglis13 commented:
^ remove configure-snapshotter subpackage

bcressey commented on packages/release/configure-snapshotter.service:
```diff
@@ -0,0 +1,25 @@
+# This file might not exist as users don't have to set `settings.container-runtime.snapshotter`
+EnvironmentFile=-/etc/containerd/selected-snapshotter
+EnvironmentFile=-/var/cache/containerd/active-snapshotter
+# Skip cleanup if selected snapshotter is empty
+ExecCondition=[ -n "${SELECTED_SNAPSHOTTER}" ]
```
For additional defense - I'm not sure how this condition would arise, but it seems safest to proceed only if we have both values and we know they're different:
```suggestion
# Skip cleanup if either snapshotter variable is empty
ExecCondition=[ -n "${SELECTED_SNAPSHOTTER}" ]
ExecCondition=[ -n "${ACTIVE_SNAPSHOTTER}" ]
```

ginglis13 replied:
I encountered a bug in testing today where snapshotter state wasn't being cleaned up on reboot in between changes to snapshotter selection and I've narrowed it down to this change. It turns out, `ExecCondition`s open the `StandardOutput` file for writing - from the [man page for systemd exec](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html):

> For units with multiple command lines, e.g. Type=oneshot services with multiple ExecStart=, or services with ExecCondition=, ExecStartPre= or ExecStartPost=, the output file is reopened and therefore re-truncated for each command line.

These `ExecCondition`s were not being met on first boot if `$SELECTED_SNAPSHOTTER` is not set (`$ACTIVE_SNAPSHOTTER` was set via the tmpfiles, but settings allow for `settings.container-runtime.snapshotter` to be unset), resulting in `$ACTIVE_SNAPSHOTTER` never being set and `/var/cache/containerd/active-snapshotter` being overwitten. From there on, the evaluation of 

```
ExecCondition=[ -n "${ACTIVE_SNAPSHOTTER}" ]
```

will never hold.

bcressey commented on packages/release/configure-snapshotter.service:
```diff
@@ -0,0 +1,25 @@
+ExecStart=-/usr/bin/find /var/lib/containerd -mindepth 1 -delete -true
+ExecStart=echo 'ACTIVE_SNAPSHOTTER="${SELECTED_SNAPSHOTTER}"'
+
+# Skip cleanup if no change detected
+RemainAfterExit=true
```
comment doesn't apply to this directive:
```suggestion
RemainAfterExit=true
```

bcressey commented on packages/release/release.spec:
```diff
@@ -197,7 +202,7 @@ install -p -m 0644 \
   %{S:1045} %{S:1046} %{S:1047} %{S:1048} %{S:1049} \
   %{S:1060} %{S:1061} %{S:1062} %{S:1063} %{S:1064} \
-  %{S:1065} %{S:1066} %{S:1067} \
+  %{S:1065} %{S:1066} %{S:1067} %{S:1068}\
```
```suggestion
  %{S:1065} %{S:1066} %{S:1067} %{S:1068} \
```

ginglis13 commented on packages/release/release.spec:
```diff
@@ -104,6 +107,7 @@ Source1500: bootconfig-fips.conf
 
 Requires: %{_cross_os}audit
 Requires: %{_cross_os}chrony
+Requires: %{_cross_os}configure-snapshotter
```
Will remove - now that we're not providing the subpackage this will result in build failures.

ginglis13 commented:
^ drop `[skip ci]`, some fixups for configure-snapshotter.service and the release spec

ginglis13 commented on packages/release/configure-snapshotter.service:
```diff
@@ -0,0 +1,25 @@
+# Don't error if the directories don't exist.
+ExecStart=-/usr/bin/find /var/lib/soci-snapshotter -mindepth 1 -delete -true
+ExecStart=-/usr/bin/find /var/lib/containerd -mindepth 1 -delete -true
+ExecStart=echo 'ACTIVE_SNAPSHOTTER="${SELECTED_SNAPSHOTTER}"'
```
```suggestion
ExecStart=/usr/bin/echo 'ACTIVE_SNAPSHOTTER="${SELECTED_SNAPSHOTTER}"'
```

ginglis13 commented:
^ force push modifies `configure-snapshotter.service`:
* writes StandardOutput to `journal+console` rather than `/var/cache/containerd/active-snapshotter`
* Use an `ExecStartPost` to copy `/etc/containerd/selected-snapshotter` to `/var/cache/containerd/active-snapshotter`

ginglis13 commented:
^ force push fixes a bug in the previous fix where `ACTIVE_SNAPSHOTTER` wasn't being set. This reverts back to the `truncate` approach to write to the active snapshotter `EnvironmentFile`, but now runs the echo to set `ACTIVE_SNAPSHOTTER="${SELECTED_SNAPSHOTTER}"` as an `ExecStopPost` so that this is run regardless of if `ExecCondition`s are met

ginglis13 commented:
^ rebase for SDK bump + clippy lint fixes

ginglis13 commented:
^ force push adjusts to writing stdout to `file:/` rather than truncating. Testing details can be found at https://gist.github.com/ginglis13/cfc7d1d907e65d889af9b55923476e12

ginglis13 commented:
^ force push `truncate`s the active snapshotter environment file before writing.

ginglis13 commented:
^ adjust soci state path in commit message
